### PR TITLE
PassBuilder

### PR DESCRIPTION
**Pull request by panyx0718**
allow user to insert pass and set pass attribute from python.

### CODE REVIEW

File: paddle/fluid/framework/ir/pass_builder.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  std::shared_ptr<Pass> AppendPass(const std::string& pass_type);
+
+  std::shared_ptr<Pass> InsertPass(size_t idx, const std::string& pass_type);
+
+  void RemovePass(size_t idx);
+
+  std::vector<std::shared_ptr<Pass>> AllPasses() const { return passes_; }
+
+ protected:
+  std::vector<std::shared_ptr<Pass>> passes_;
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/pass_builder.h]: Why shared_ptr here? seems unique_ptr works.

**Review by Superjomn — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/ir/pass_builder.h (hunk 1 above)
Comment (panyx0718) [reply paddle/fluid/framework/ir/pass_builder.h]: PassBuilder needs to be copyable (shared with Python), so, the stuffs it hold are shared_ptr as well.

**Comment by chengduoZH** [comment]
Is it necessary that letting users add pass before execution?

File: paddle/fluid/framework/details/build_strategy.cc [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    if (!strategy_.debug_graphviz_path_.empty()) {
+      auto viz_pass = AppendPass("graph_viz_pass");
+      const std::string graph_path = string::Sprintf(
+          "%s%s", strategy_.debug_graphviz_path_.c_str(), "_original_graph");
+      viz_pass->Set<std::string>("graph_viz_path", new std::string(graph_path));
+    }
+
+    // Add op fusion.
+    if (strategy.fuse_elewise_add_act_ops_) {
+      auto fuse_elewise_add_act_pass = AppendPass("fuse_elewise_add_act_pass");
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/build_strategy.cc:40]: `fuse_elewise_add_act_pass` is unused here.

Comment (panyx0718) [reply paddle/fluid/framework/details/build_strategy.cc:40]: Right, in this method, we only assemble passes. Passes are applied in Apply() method below

**Review by chengduoZH — APPROVED** [review]
LGTM

**Pull request merged**